01-data-structures/05-hashes-part-2/separate_chaining/separate_chaining.rb

```diff
@@ -1,33 +1,85 @@
+require_relative 'node'
 require_relative 'linked_list'
 
 class SeparateChaining
   attr_reader :max_load_factor
 
   def initialize(size)
     @max_load_factor = 0.7
+    @items = Array.new(size)
   end
 
   def []=(key, value)
+    node = Node.new(key, value)
+    i = index(key, self.size)
+    if @items[i].nil?
+      item = LinkedList.new
+      item.add_to_tail(node)
+      @items[i] = item
+    elsif @items[i].size < 5
+      @items[i].add_to_tail(node)
+    else
+      resize
+      self[key] = value
+    end
+    if self.load_factor > @max_load_factor
+      resize
+    end
   end
 
   def [](key)
+    i = index(key, self.size)
+    if !@items[i].nil?
+      node = @items[i].find_node(key)
+      if !node.value.nil?
+        return node.value
+      end
+    end
   end
 
   # Returns a unique, deterministically reproducible index into an array
   # We are hashing based on strings, let's use the ascii value of each string as
   # a starting point.
   def index(key, size)
+    key.sum % size
   end
 
   # Calculate the current load factor
+  # Load factor is the number of elements in a hash table divided by the number of buckets it has allocated.
   def load_factor
+    x = 0.0
+    for i in 0..self.size-1
+      if !@items[i].nil?
+        x += @items[i].size
+      end
+    end
+    x / self.size
```

Comment:
This can be simplified by maintaining a property like `@node_count` and increment it each time a node is added. Of course, you would have to make sure to reset this during each resize.